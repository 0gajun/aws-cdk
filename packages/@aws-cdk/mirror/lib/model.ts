export type NodeMetadata = { [name: string]: any };
export type NodeLinks = Link[];

export interface Link {
  sourcePath: string;
  targetPath: string;
  attribute?: string;
}

export interface Node {
  id: string;
  path: string;
  children?: Node[];
  metadata?: NodeMetadata;
  links?: NodeLinks;
}
