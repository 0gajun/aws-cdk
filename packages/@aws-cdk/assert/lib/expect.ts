import cdk = require('@aws-cdk/core');
import api = require('@aws-cdk/cx-api');
import { StackInspector } from './inspector';
import { SynthUtils } from './synth-utils';

export function expect(stack: api.CloudFormationStackArtifact | cdk.Stack, skipValidation = false): StackInspector {
  if (stack instanceof api.CloudFormationStackArtifact) {
    return new StackInspector(stack);
  }
  return new StackInspector(SynthUtils.synthesize(stack, { skipValidation }), stack);
}
