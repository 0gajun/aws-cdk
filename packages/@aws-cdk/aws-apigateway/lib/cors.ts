export interface CorsOptions {
  /**
   * @default 204
   */
  readonly statusCode?: number;

  readonly allowOrigin: string;

  /**
   * @default
   */
  readonly allowHeaders?: string[];

  /**
   * @default
   */
  readonly allowMethods?: string[];

  /**
   * @default false
   */
  readonly allowCredentials?: boolean;
}

export class Cors {
  // TODO: dedup with utils.ts/ALLOWED_METHODS
  public static readonly ALL_METHODS = [ 'OPTIONS', 'GET', 'PUT', 'POST', 'DELETE', 'PATCH', 'HEAD' ];
  public static readonly DEFAULT_HEADERS = [ 'Content-Type', 'X-Amz-Date', 'Authorization', 'X-Api-Key', 'X-Amz-Security-Token', 'X-Amz-User-Agent' ];
}