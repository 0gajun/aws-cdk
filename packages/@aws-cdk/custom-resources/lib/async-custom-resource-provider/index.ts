// tslint:disable:no-console
import SDK = require('aws-sdk');
import { CompleteResourceEvent, ResourceEvent } from './api';
import { MyResourceHandler } from './demo';

const sfn = new SDK.StepFunctions();
const asyncHandler = new MyResourceHandler();

async function failOnError(event: ResourceEvent, block: () => Promise<any>) {
  try {
    return await block();
  } catch (e) {
    if (e instanceof Retry) {
      throw e; // state machine will retry
    }

    // this is an actual error, fail the activity altogether and exist.
    await respond(event, 'FAILED', e.message);

    return {
      error: e.message
    };
  }
}

exports.begin = async (event: ResourceEvent) => {
  console.log(JSON.stringify(event));

  await failOnError(event, async () => {
    const result = asyncHandler.begin(event);

    if (event.PhysicalResourceId && result.PhysicalResourceId !== event.PhysicalResourceId) {
      throw new Error(`Physical resource ID cannot be changed from ${event.PhysicalResourceId} to ${result.PhysicalResourceId}`);
    }

    const input: CompleteResourceEvent = {
      ...event,
      PhysicalResourceId: result.PhysicalResourceId,
      Data: result.Data,
    };

    const stateMachineArn = process.env.STATE_MACHINE_ARN;
    if (!stateMachineArn) {
      throw new Error(`STATE_MACHINE_ARN is not defined`);
    }

    const req: AWS.StepFunctions.StartExecutionInput = {
      stateMachineArn,
      name: event.RequestId,
      input: JSON.stringify(input),
    };

    return await sfn.startExecution(req).promise();
  });
};

// invoked a few times until `complete` is true or until it times out.
exports.complete = async (event: CompleteResourceEvent) => {
  console.log({ complete: event });

  return await failOnError(event, async () => {

    const result = asyncHandler.complete(event);
    if (!result.complete) {
      throw new Retry(JSON.stringify(event));
    }

    // done, respond with success
    await respond(event, 'SUCCESS', 'Done');
  });
};

// invoked when completion retries are exhaused.
exports.timeout = async (event: any) => {
  console.log(JSON.stringify(event));
  const e: ResourceEvent = JSON.parse(JSON.parse(event.Cause).errorMessage);
  await respond(e, 'FAILED', 'Operation timed out');
};

class Retry extends Error { }

function respond(event: ResourceEvent, responseStatus: 'SUCCESS' | 'FAILED', reason: string) {
  const responseBody = JSON.stringify({
    Status: responseStatus,
    Reason: reason,
    StackId: event.StackId,
    RequestId: event.RequestId,
    PhysicalResourceId: event.PhysicalResourceId,
    LogicalResourceId: event.LogicalResourceId,
    NoEcho: false,
    Data: event.Data
  });

  console.log('Responding', responseBody);

  const parsedUrl = require('url').parse(event.ResponseURL);
  const requestOptions = {
    hostname: parsedUrl.hostname,
    path: parsedUrl.path,
    method: 'PUT',
    headers: { 'content-type': '', 'content-length': responseBody.length }
  };

  return new Promise((resolve, reject) => {
    try {
      const request = require('https').request(requestOptions, resolve);
      request.on('error', reject);
      request.write(responseBody);
      request.end();
      resolve();
    } catch (e) {
      reject(e);
    }
  });
}
