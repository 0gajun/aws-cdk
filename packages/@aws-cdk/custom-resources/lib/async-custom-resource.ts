import cfn = require('@aws-cdk/aws-cloudformation');
import lambda = require('@aws-cdk/aws-lambda');
import sfn = require('@aws-cdk/aws-stepfunctions');
import tasks = require('@aws-cdk/aws-stepfunctions-tasks');
import { Construct, Duration, RemovalPolicy, Stack } from '@aws-cdk/core';
import path = require('path');

export interface AsyncCustomResourceProps {
  readonly uuid: string;
  readonly properties?: { [name: string]: any };
  readonly resourceType?: string;
  readonly removalPolicy?: RemovalPolicy;

  /**
   * How many seconds to wait initially before retrying
   *
   * @default Duration.seconds(1)
   */
  readonly interval?: Duration;

  /**
   * How many times to retry this particular error.
   *
   * May be 0 to disable retry for specific errors (in case you have
   * a catch-all retry policy).
   *
   * @default 3
   */
  readonly maxAttempts?: number;

  /**
   * Multiplication for how much longer the wait interval gets on every retry
   *
   * @default 2
   */
  readonly backoffRate?: number;
}

export class AsyncCustomResource extends cfn.CustomResource {
  constructor(scope: Construct, id: string, props: AsyncCustomResourceProps) {
    const handlerStack = getCreateHandlerStack(scope, props);

    super(scope, id, {
      provider: cfn.CustomResourceProvider.lambda(handlerStack.begin),
      resourceType: props.resourceType,
      properties: props.properties,
      removalPolicy: props.removalPolicy
    });
  }
}

function getCreateHandlerStack(scope: Construct, props: AsyncCustomResourceProps) {
  const stack = Stack.of(scope);
  const id = props.uuid;
  const exists = stack.node.tryFindChild(id) as HandlerStack;
  if (exists) {
    return exists;
  }

  return new HandlerStack(stack, id, props);
}

const HANDLER_PATH = path.join(__dirname, 'async-custom-resource-provider');

class HandlerStack extends Stack {
  public readonly begin: lambda.Function;

  constructor(scope: Construct, id: string, props: AsyncCustomResourceProps) {
    super(scope, id, { stackName: id });

    const completeTask = this.taskForEntrypoint('complete');
    completeTask.addCatch(this.taskForEntrypoint('timeout'));
    completeTask.addRetry({
      maxAttempts: props.maxAttempts,
      interval: props.interval,
      backoffRate: props.backoffRate,
    });

    const stateMachine = new sfn.StateMachine(this, 'state-machine', {
      definition: completeTask
    });

    this.begin = this.handlerForEntrypoint('begin');
    this.begin.addEnvironment('STATE_MACHINE_ARN', stateMachine.stateMachineArn);
    stateMachine.grantStartExecution(this.begin);
  }

  private handlerForEntrypoint(entrypoint: string) {
    return new lambda.Function(this, `${entrypoint}-handler`, {
      code: lambda.Code.fromAsset(HANDLER_PATH),
      runtime: lambda.Runtime.NODEJS_10_X,
      handler: `index.${entrypoint}`
    });
  }

  private taskForEntrypoint(entrypoint: string) {
    return new sfn.Task(this, `${entrypoint}-task`, {
      task: new tasks.InvokeFunction(this.handlerForEntrypoint(entrypoint)),
    });
  }
}
