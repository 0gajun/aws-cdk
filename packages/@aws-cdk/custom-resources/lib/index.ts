export * from './aws-custom-resource';
export * from './async-custom-resource';