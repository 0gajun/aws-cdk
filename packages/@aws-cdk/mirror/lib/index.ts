export * from './mirror';
export * from './model';
