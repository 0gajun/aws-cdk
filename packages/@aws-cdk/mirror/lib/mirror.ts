import cdk = require('@aws-cdk/cdk');
import fs = require('fs');
import { Link, Node, NodeLinks, NodeMetadata } from './model';

/**
 * @experimental
 */
export class Mirror extends cdk.Construct {
  public static readonly FileName = 'mirror.json';

  constructor(scope: cdk.Construct, id: string) {
    super(scope, id);
  }

  protected synthesize(session: cdk.ISynthesisSession) {
    const lookup: { [path: string]: Node } = { };

    const visit = (construct: cdk.IConstruct): Node => {
      const children = construct.node.children.map(visit);
      const metadata = this.renderMetadata(construct);
      const node: Node = {
        id: construct.node.id || 'App',
        path: construct.node.path,
        children: children.length === 0 ? undefined : children,
        metadata: Object.keys(metadata).length === 0 ? undefined : metadata,
      };

      lookup[node.path] = node;

      return node;
    };

    const root = visit(this.node.root);

    // add links to all nodes (in both sides of the edge).
    for (const link of this.findLinks()) {
      const sourceNode = lookup[link.sourcePath];
      const targetNode = lookup[link.targetPath];
      sourceNode.links = [ ...sourceNode.links || [], link ];
      targetNode.links = [ ...targetNode.links || [], link ];
    }

    // temporary workaround until toolkit actually retains the cloud assembly
    fs.writeFileSync(Mirror.FileName, JSON.stringify(root, undefined, 2));

    session.store.writeJson(Mirror.FileName, root);
  }

  private renderMetadata(construct: cdk.IConstruct): NodeMetadata {
    if (cdk.Resource.isResource(construct)) {
      return {
        resourceType: construct.resourceType,
        logicalId: construct.node.resolve(construct.logicalId)
      };
    }

    return {};
  }

  private findLinks(): NodeLinks {
    const links = new Array<Link>();

    for (const edge of this.node.root.node.findReferences()) {
      if (cdk.GetAtt.isGetAtt(edge.reference)) {
        links.push({
          sourcePath: edge.source.node.path,
          targetPath: edge.target.node.path,
          attribute: edge.reference.attributeName
        });
      } else if (cdk.Ref.isRef(edge.reference)) {
        links.push({
          sourcePath: edge.source.node.path,
          targetPath: edge.target.node.path
        });
      }
    }

    return links;
  }
}
