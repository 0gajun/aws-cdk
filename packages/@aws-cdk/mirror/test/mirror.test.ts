import { App, Construct, Resource, Stack } from '@aws-cdk/cdk';
import { Mirror } from '../lib';

test('hey', () => {
  const app = new App();

  const stack = new Stack(app, 'myapp');
  const group1 = new Construct(stack, 'group1');
  const group2 = new Construct(stack, 'group2');

  const resource1 = new Resource(group1, 'group1-resource1', { type: 'my-resource-type' });
  new Resource(group1, 'group1-resource2', { type: 'my-resource-type-2' });

  new MyResource(group2, 'group2-resource3', {
    resource: resource1
  });

  new Mirror(app, 'mirror');

  const session = app.run();
  const output = session.store.readJson(Mirror.FileName);

  expect(output).toMatchSnapshot();
});

interface MyResourceProps {
  resource: Resource
}

class MyResource extends Resource {
  constructor(scope: Construct, id: string, props: MyResourceProps) {
    super(scope, id, {
      type: 'resource-type',
      properties: {
        SomeRef: props.resource.ref,
        SomeAttribute: props.resource.getAtt('Boom').toString()
      }
    });
  }
}
