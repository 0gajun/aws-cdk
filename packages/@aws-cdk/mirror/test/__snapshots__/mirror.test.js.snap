// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hey 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": "group1-resource1",
              "links": Array [
                Object {
                  "attribute": "Boom",
                  "sourcePath": "myapp/group2/group2-resource3",
                  "targetPath": "myapp/group1/group1-resource1",
                },
              ],
              "metadata": Object {
                "logicalId": "group1group1resource1C98E375E",
                "resourceType": "my-resource-type",
              },
              "path": "myapp/group1/group1-resource1",
            },
            Object {
              "id": "group1-resource2",
              "metadata": Object {
                "logicalId": "group1group1resource2E9646EF4",
                "resourceType": "my-resource-type-2",
              },
              "path": "myapp/group1/group1-resource2",
            },
          ],
          "id": "group1",
          "path": "myapp/group1",
        },
        Object {
          "children": Array [
            Object {
              "id": "group2-resource3",
              "links": Array [
                Object {
                  "attribute": "Boom",
                  "sourcePath": "myapp/group2/group2-resource3",
                  "targetPath": "myapp/group1/group1-resource1",
                },
              ],
              "metadata": Object {
                "logicalId": "group2group2resource385A87E2A",
                "resourceType": "resource-type",
              },
              "path": "myapp/group2/group2-resource3",
            },
          ],
          "id": "group2",
          "path": "myapp/group2",
        },
      ],
      "id": "myapp",
      "path": "myapp",
    },
    Object {
      "id": "mirror",
      "path": "mirror",
    },
  ],
  "id": "App",
  "path": "",
}
`;
